Add support for multipart downloads for module-format Worker scripts

sodabrew opened the pull request:
## Description

Since #1010 allows uploading a Worker script in module format, calling DownloadWorker on the same script will return a multipart/form-data response.

### Asks

1. _Update: still need this; for the moment I'm assuming that any multipart worker is in module format._ Cloudflare API for download a worker should add a Content-Type header to each multipart/form-data section to specify whether the script in that section is application/javascript or application/javascript+module.
2. _Update: took this approach, please review!_ Guidance on a deep change: `makeRequest...` et al should return a content-type header not just the body.
3. _Update: didn't like this approach:_ ~Guidance on a workaround: `makeRequest...` et al continue to just return the response body, so consumers assume the format, or detect the format with a heuristic. For example, `DownloadWorker` can check the first two bytes and if they are `--` then assume the rest of that line is the multipart boundary string and parse the body accordingly. This is so gross.~

### Details

This approach digs into the `makeRequest...` family (`makeRequestWithAuthTypeAndHeaders` is the full implementation) and adds support to return the first body of a multipart response. `makeRequest`, et al, does not return the mime-type of the response. Consuming functions mostly assume it is application/json, except DownloadWorker which implicitly assumes it's application/javascript.

While I've solved the initial problem with #1010 that you can upload a module-format worker but if you try to download it, the DownloadWorker method return raw multipart/form-data, I'm stuck on the next problem: the Module true/false status is _implied_ by whether the download script API returns a javascript content-type or a multipart content-type.

There's no further clue because the Cloudflare API for downloading a worker only gives a Content-Disposition header on its response for a module worker:
```
Content-Type: multipart/form-data; boundary=mimeboundary

--mimeboundary
Content-Disposition: form-data; name="worker.js"

export default { etc etc etc }
--mimeboundary
```
vs. non-module format:
```
Content-Type: application/json

addEventListener( etc etc )
```

This isn't consistent with **uploading** a module worker, where the Content-Type of the mime-part must be set to `application/javascript+module` like this:
```
--mimeboundary
Content-Disposition: form-data; name="worker.js"
Content-Type: application/javascript+module

export default { etc etc etc }
--mimeboundary
```

Sidenote: why was the method `UploadWorkerWithBindings` added? The bindings are in a param structure, isn't the point to add new params without creating a new method? And in fact if you call UploadWorker with Module-format worker, it shunts to UploadWorkerWithBindings! The salient difference is whether the upload is in multipart format or a direct script. The code path I can see that isn't in multipart format is a non-module worker without bindings (the original product, so probably also the most common case). _Per manual testing, the Cloudflare API will accept a multipart/form-data with a plain script as its body -- that is, **UploadWorker** is sufficient for all cases and **UploadWorkerWithBindings** is duplicative._

## Has your change been tested?

_**Works**_ but creates an awkward roundtrip problem in the current iteration:

1. `UploadWorker` with Module: true because it's a module script.
```
        scriptFile, err := os.ReadFile("cool-script.js")
        scriptText := string(scriptFile)

        UploadWorker(ctx,
                &cloudflare.WorkerRequestParams{
                        ScriptName: "modtest",
                },
                &cloudflare.WorkerScriptParams{
                        Script: scriptText,
                        Module: true,
                },
        )
```
4. `DownloadWorker` with the script name and returns the script body. But what format is it?
```
        scriptText = DownloadWorker(ctx,
                &cloudflare.WorkerRequestParams{
                        ScriptName: "modtest",
                },
        )
```
5. `UploadWorker` ... with Module true or false? There's no indication from DownloadWorker which to use.
```
        strings.Replace(scriptText, "foo", "bar", -1)

        UploadWorker(ctx,
                &cloudflare.WorkerRequestParams{
                        ScriptName: "modtest",
                },
                &cloudflare.WorkerScriptParams{
                        Script: scriptText,
                        Module: true,   <---- ?????????????????                },
        )
```


## Screenshots (if appropriate):

## Types of changes

What sort of change does your code introduce/modify?

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented (api.cloudflare.com or developers.cloudflare.com) and stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

sodabrew committed:
Add support for multipart downloads for module-format Worker scripts

sodabrew commented on cloudflare.go:
```diff
@@ -265,6 +267,17 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
 			continue
+		}
+
+		// Successful request, let's read the body and return the first part
```
Need to fix this comment, as the code is returning the _last_ part if there are multiple parts.

sodabrew commented on cloudflare.go:
```diff
@@ -273,13 +286,14 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
 			break
 		}
-	}
+	} // for loop
```
I added this comment to help understand where in the control flow the reader block was taking place. I actually think the body reading block should move out on control level -- that is, a successful API response breaks the for loop, then read the body.

sodabrew commented on workers_test.go:
```diff
@@ -179,15 +179,38 @@ const (
 
 var (
 	successResponse               = Response{Success: true, Errors: []ResponseInfo{}, Messages: []ResponseInfo{}}
-	workerScript                  = "addEventListener('fetch', event => {\n    event.passThroughOnException()\nevent.respondWith(handleRequest(event.request))\n})\n\nasync function handleRequest(request) {\n    return fetch(request)\n}"
```
Way easier to read using backticks than embedded newlines, IMO. No functional change here.

sodabrew commented on workers_test.go:
```diff
@@ -384,13 +407,33 @@ func TestWorkers_DownloadWorkerWithNameErrorsWithoutAccountId(t *testing.T) {
 	mux.HandleFunc("/accounts/foo/workers/scripts", func(w http.ResponseWriter, r *http.Request) {
 		assert.Equal(t, http.MethodGet, r.Method, "Expected method 'GET', got %s", r.Method)
-		w.Header().Set("content-type", "application-json")
+		w.Header().Set("content-type", "application/json")
```
`makeRequestWithAuthTypeAndHeaders` doesn't look at the content-type of the response, so it never flagged the incorrect "`application-json`"[sic] content-type before. I think it would be helpful to be stricter about only parsing JSON responses if the response content-type is set appropriately.

sodabrew committed:
Workers script test data whitespace for clarity

sodabrew commented:
@jacobbednarz thanks for quickly merging the two test fixits that I pulled out of this PR into #1048 and #1049. I wrote a massive update to the OP with examples and problems of implied content-type. Looking forward to your comments!

jacobbednarz commented on cloudflare.go:
```diff
@@ -341,6 +339,32 @@ func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, u
+		}
+	}
+
+	if api.Debug {
```
this probably shouldn't move as the debug is intended to also allow seeing non-200 HTTP responses to help problem diagnosis.

sodabrew replied:
Moved this back where it was and moved the multipart parsing to DownloadWorker itself. Check out this new approach and let me know what you think!

sodabrew replied:
ps ... what's going on in cloudflare_experimental.go? It looks like a subtly different API client implementation and is only used in the Zone code.

jacobbednarz commented:
this PR has become quite significantly larger than the intended scope and i'm afraid, i won't be merging it as is with such a large refactor of the core methods. leading into your next question, the experimental client (see https://github.com/cloudflare/cloudflare-go/blob/master/docs/experimental.md) will focus on these larger, sweeping improvements for the next major release. we'll be intentionally shaking away alot of the older library functionality that we have now which would be a better place for type of changes you're proposing here. i wouldn't recommend opening PRs for it at this stage, but, just be aware that it exists.

for now, if you want this functionality, we need to do it within the confines of the library API as it stands todya.

sodabrew commented:
Ok, thanks for that feedback. I've adjusted the scope of changes to be a lot narrower, please take another look and let me know.

jacobbednarz committed:
make APIResponse match more like http.Response

jacobbednarz commented:
i've gone ahead and updated `APIResponse` to be more `http.Response` like as that will be the direction for the experimental client when we start porting functionality to it. this looks fine in the tests but do you want to confirm locally that this meets what you're after here before we merge?

jacobbednarz commented on cloudflare.go:
```diff
@@ -176,21 +176,41 @@ func (api *API) makeRequestContextWithHeaders(ctx context.Context, method, uri s
+}
+
+// Use this method if an API response can have different Content-Type headers and different body formats.
+func (api *API) makeRequestContextWithHeadersComplete(ctx context.Context, method, uri string, params interface{}, headers http.Header) (*APIResponse, error) {
```
i don't love the name here however, i suspect this will be cleaned up in the experimental client so i'm not too phase but will want to keep it's usage _extremely_ limited until then.

sodabrew commented:
Looks great! I've rebased and squashed down to three commits: the core change I arrived at, your edits thereupon, and worker test script whitespace changes. You might need to repush your last commit with your own signing key (I see an "Unverified" note) -- or I could squash your final commit into mine with a Co-authored-by?

jacobbednarz commented:
done! thanks for the PR and getting it over the line. i've made some notes internally around the issues here with headers and the request methods not exposing everything which hopefully we can address nicely in the experimental client later on.
